edk2toollib/uefi/edk2/build_objects/dsc.py

```diff
@@ -0,0 +1,455 @@
+            return False
+        arch = "*" if (self.arch == "COMMON" or self.arch == "DEFAULT") else self.arch
+        arch2 = "*" if (other.arch == "COMMON" or other.arch == "DEFAULT") else other.arch
+        return self.module_type == other.module_type or arch == arch2
+
+    def __repr__(self):
+        attributes = f".{self.arch}.{self.module_type}"
+        return attributes
+
+    @classmethod
+    def IsValidModuleType(cls, name):
+        return name in cls.dsc_module_types
+
+
+class dsc_buildoption_section_type(dsc_section_type):
+    '''
+    [BuildOptions.$(arch).CodeBase.Edk2ModuleType]
+    [BuildOptions.$(arch).CodeBase]
+    [BuildOptions.common.CodeBase]
+    [BuildOptions.$(arch)]
+    [BuildOptions.common]
+    [BuildOptions]
+    '''
+
+    def __init__(self, arch="common", codebase="common", module_type="common"):
+        super().__init__(arch, module_type)
+        self.codebase = codebase.upper().strip()
+        if not self.IsValidCodeBase(self.codebase):
+            raise ValueError(f"{codebase} is not a valid codebase type")
+
+    @classmethod
+    def IsValidCodeBase(cls, codebase):
+        return codebase in ["COMMON", "EDK", "EDKII"]
+
+    def __hash__(self):
+        return hash((super().__hash__(), self.codebase))
+
+    def __eq__(self, other):
+        if type(other) is not dsc_buildoption_section_type:
+            return False
+        return other.module_type == self.module_type and other.codebase == self.codebase and other.module_type == self.module_type
+
+    def __repr__(self):
+        return f"{self.arch}.{self.codebase}.{self.module_type}"
+
+
+dsc_pcd_types = ["FEATUREFLAG", "PATCHABLEINMODULE", "FIXEDATBUILD", "DYNAMIC", "DYNAMICEX",
+                 "DYNAMICDEFAULT", "DYNAMICHII", "DYNAMICVPD", "DYNAMICEXHII", "DYNAMICEXVPD"]
+
+
+class dsc_pcd_section_type():
+    def __init__(self, pcdtype, arch="common", sku="DEFAULT"):
+        self.arch = arch.upper().strip()
+        self.pcd_type = pcdtype.upper().strip()
+        if self.pcd_type not in dsc_pcd_types:
+            raise ValueError(f"{pcdtype} is not a proper PCD type")
+        self.sku = sku.upper()
+
+    def __hash__(self):
+        return hash((self.arch, self.pcd_type))
+
+    def __repr__(self):
+        return f"Pcds{self.pcd_type}.{self.arch}.{self.sku}"
+
+    def __eq__(self, other):
+        if type(other) is not dsc_pcd_section_type:
+            return False
+        return self.pcd_type == other.pcd_type and self.arch == other.arch
+
+
+class dsc_pcd_component_type(dsc_pcd_section_type):
+    ''' This class is uses to define the PCD type inside a component '''
+
+    def __init__(self, pcdtype):
+        super().__init__(pcdtype)
+
+    def __repr__(self):
+        return f"Pcds{self.pcd_type}"
+
+    def __hash__(self):
+        return hash(self.pcd_type)
+
+    def __eq__(self, other):
+        if type(other) is not dsc_pcd_component_type:
+            return False
+        return self.pcd_type == other.pcd_type
+
```

Comment:
why there are only section type class for pcd and buildoption?

Reply:
LibraryClass uses the dsc_section_type type to bisect it.